ERCS/erc-7677.md

```diff
@@ -0,0 +1,350 @@
+  },
+  "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789",
+  "0x2105",
+  {
+    // Illustrative context field. These should be defined by service providers.
+    "policyId": "962b252c-a726-4a37-8d86-333ce0a07299"
+  }
+]
+```
+
+###### `pm_getPaymasterStubData` Example Return Value
+
+Paymaster services MUST detect which entrypoint version the account is using and return the correct fields.
+
+For example, if using entrypoint v0.6:
+
+```json
+{
+  "paymasterAndData": "0x..."
+}
+```
+
+If using entrypoint v0.7:
+
+```json
+{
+  "paymaster": "0x...",
+  "paymasterData": "0x..."
+}
+```
+
+If using entrypoint v0.7, with paymaster gas estimates:
+
+```json
+{
+  "paymaster": "0x...",
+  "paymasterData": "0x...",
+  "paymasterVerificationGasLimit": "0x...",
+  "paymasterPostOpGasLimit": "0x..."
+}
+```
+
+#### `pm_getPaymasterData`
+
+Returns values to be used in paymaster-related fields of a signed user operation. These are not stub values and will be used during user operation submission to a bundler. Similar to `pm_getPaymasterStubData`, accepts an unsigned user operation, entrypoint address, chain id, and a context object.
+
+This method also returns a `sponsor` object with `name` and `icon` fields. The `name` field is the name of the party sponsoring the transaction, and the `icon` field is a URI pointing to an image. Wallet developers MAY choose to display sponsor information to users. The `icon` string MUST be a data URI as defined in [RFC-2397]. The image SHOULD be a square with 96x96px minimum resolution. The image format is RECOMMENDED to be either lossless or vector based such as PNG, WebP or SVG to make the image easy to render on the wallet. Since SVG images can execute Javascript, wallets MUST render SVG images using the `<img>` tag to ensure no untrusted Javascript execution can occur.
+
+##### `pm_getPaymasterData` RPC Specification
+
+Note that the user operation parameter does not include a signature, as the user signs after all other fields are populated.
+
+```typescript
+// [userOp, entryPoint, chainId, context]
+type GetPaymasterDataParams = [
+  // Below is specific to Entrypoint v0.6 but this API can be used with other entrypoint versions too
+  {
+    sender: `0x${string}`;
+    nonce: `0x${string}`;
+    initCode: `0x${string}`;
+    callData: `0x${string}`;
+    callGasLimit: `0x${string}`;
+    verificationGasLimit: `0x${string}`;
+    preVerificationGas: `0x${string}`;
+    maxFeePerGas: `0x${string}`;
+    maxPriorityFeePerGas: `0x${string}`;
+  }, // userOp
+  `0x${string}`, // Entrypoint
+  `0x${string}`, // Chain ID
+  Record<string, any> // Context
+];
+
+type GetPaymasterDataResult = {
+  sponsor?: { name: string; icon: string } // Sponsor info
+  paymaster?: string // Paymaster address (entrypoint v0.7)
+  paymasterData?: string; // Paymaster data (entrypoint v0.7)
+  paymasterAndData?: string; // Paymaster and data (entrypoint v0.6)
+}
+```
+
+###### `pm_getPaymasterData` Example Parameters
```

Comment:
Paymaster services are supposed to detect which account implementation is it from the sender's address?
otherwise how would one decode data based on policy-id?

Reply:
detecting the account implementation is not a strict requirement of paymaster services, and it is ultimately up to paymaster services to decide how to validate policy ids or whatever other abstractions they might have.

you could reliably do this without needing to know the implementation though by simulating the incoming user op, no need to decode the calldata (although you could if you know the implementation).

Reply:
Shouldn't this be a requirement as it is mentioned that 
`Paymaster services MUST detect which entrypoint version the account is using and return the correct fields.`
how is a paymaster supposed to detect this without knowing the account implementation?
Afaik all smart accounts store the supported entry point on the smart account in a simple variable and reading that might be the best way to know this information.
To make that possible I think the paymaster services should have some knowledge about the implementation

Reply:
paymaster knows the version from the entrypoint address (in any case, a paymaster support explicit entrypoint instance, not "any".
it would still be nice if at the API level a "version" field is added..